File: Q99-PlaceQueen.cpp
//
//  Q99-PlaceQueen.cpp
//  Algorithm-Linux
//
//  Created by Sanqiang Zhao on 12/15/12.
//  Copyright (c) 2012 Sanqiang Zhao. All rights reserved.
//
/*
 vector<int*> so = placeQueen(8, 0);
 for (int i = 0; i < so.size(); ++i) {
 printQueen(so[i], 8, 8);
 cout<<"======"<<endl;
 }
*/
#include "Q99-PlaceQueen.h"
vector<int*> solutions;
int * check;
vector<int*> placeQueen(int edge, int row)
{
    if (row == edge) {
        //printQueen(check, edge, edge);
        int * clone = new int[edge];
        for (int c = 0; c < edge; ++c) {
            clone[c] = check[c];
        }
        solutions.push_back(clone);
        return solutions;
    }
    
    if (row == 0) {
        check = new int[edge];
    }
    int i = 0;
    for (; i < edge; ++i) {
        if (checkQueen(check, row, i, edge)) {
            check[row] = i;
            placeQueen(edge, row+1);
        }
    }
    
    return solutions;
}

bool checkQueen(int check[], int row, int col, int edge)
{
    int r = 0;
    for (; r < row; ++r) {
        if (check[r] == col || abs(r - row) == abs(check[r] - col)) {
            return false;
        }
    }
    return true;
}


void printQueen(int arr[], int length, int edge)
{
    int i = 0;
    for (; i < length; ++i) {
        int pos = arr[i];
        int j = 0;
        for (; j < pos; ++j) {
            std::cout<<"0";
        }
        std::cout<<"1";
        for (j = j + 1; j < edge; ++j) {
            std::cout<<"0";
        }
        std::cout<<std::endl;
    }

}

File: Algorithm-Java/src/CareerCup/RecursiveAndDP/Q9_9_PlaceQueen.java
/**
 * Sanqiang Zhao Www.131X.Com Jan 5, 2013
 */
package CareerCup.RecursiveAndDP;

public class Q9_9_PlaceQueen {

    final int GRID_SIZE = 8;

    public void place() {
        place(0, new int[GRID_SIZE]);
    }

    public void place(int row, int[] chess) {
        if (row == GRID_SIZE) {
            print(chess);
            return;
        }

        for (int i = 0; i < GRID_SIZE; i++) {
            chess[row] = i;
            if (checkValid(row, chess)) {
                place(row + 1, chess);
            }
        }
    }

    private boolean checkValid(int row, int[] chess) {
        for (int i = 0; i < row; i++) {
            int w = Math.abs(chess[i] - chess[row]);
            int h = Math.abs(i - row);
            if (w == 0 || w == h) {
                return false;
            }
        }
        return true;
    }

    private void print(int[] chess) {
        for (int i = 0; i < GRID_SIZE; i++) {
            int col = chess[i];
            int j = 0;
            for (; j < col; j++) {
                System.out.print("_");
            }
            System.out.print("*");
            for (j = col + 1; j < GRID_SIZE; j++) {
                System.out.print("_");
            }
            System.out.println();
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Q9_9_PlaceQueen q99 = new Q9_9_PlaceQueen();
        q99.place();
    }
}
